Add dynamical phase transitions tutorial

jpacold opened the pull request:
## Description

Wrote a tutorial notebook to close #1795.

The tutorial follows [this paper](https://arxiv.org/abs/2211.08318), roughly reproducing Fig. 2 (e) and (f). The main difference is that I did much coarser simulations, to get the whole notebook to run in ~5 minutes on my machine.

## Question

I initially tried using the high-level functions [`execute_with_zne`](https://mitiq.readthedocs.io/en/latest/apidoc.html#mitiq.zne.zne.execute_with_zne) and [`mitigate_executor`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.zne.zne.mitigate_executor) for the ZNE step, but got a circuit conversion error (I'm trying to put together a smaller example to reproduce this). I ended up using `RichardsonFactory.extrapolate` to do the extrapolation and am not sure I did this correctly.

---

### License

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

cosenal commented:
@jpacold Awesome to see that you managed to reproduce those results!

Before we review it, I kindly ask you to look into the build failures. 

1. `build / validate` should be easy to fix if you run those command in the [Style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines) section of Mitiq docs.
2. For the `docs-build` one, you want to add your tutorial to the toctree
```
 /home/runner/work/mitiq/mitiq/docs/source/examples/dynamical_phase_transitions_zne.md: WARNING: document isn't included in any toctree
 ```
 3. The `build / test` one is probably a flaky test that fails doing to an unlucky random event (somehow the noisy case ends up having a perfect value). Feel free to look into this, or propose a solution, but it looks outside the scope of your PR.

purva-thakre commented:
> For the docs-build one, you want to add your tutorial to the toctree

To support Alessandro's comment on this, the tutorial should be added to `docs/source/examples/examples.md`.

jpacold commented:
Thanks for the advice! It looks like there is still a test failure but the docs build should be fixed.

cosenal commented:
Thanks, @jpacold. Regarding your question in the PR description, most definitely a small non-working example would help to understand what the issue is.

cosenal commented on docs/source/examples/dynamical_phase_transitions_zne.md:
```diff
@@ -0,0 +1,352 @@
++++
+
+# Circuit definition
+
```
Ideally we want those tutorials to be self-contained, and explain the idea behind the paper without referencing figures and equations from there.

cosenal commented on docs/source/examples/dynamical_phase_transitions_zne.md:
```diff
@@ -0,0 +1,352 @@
+
+# Introduction
+
+This tutorial replicates some of the results from Y. Javanmard et al., ["Quantum simulation of dynamical phase transitions in noisy quantum devices"](https://arxiv.org/abs/2211.08318). We build a circuit that simulates the the time-evolution of a transverse-field Ising model (Eq. 1) after a quench at time $t = 0$, then run ideal, noisy, and noise-mitigated simulations of the circuit.
```
s/the the/the

cosenal reviewed (commented):
some preliminary comments

jpacold commented:
- Fixed heading levels

- Substantially expanded the introduction to describe the circuit without references to the paper

- Changed the title. The results in the paper that are specifically about dynamical phase transitions are outside the current scope of this tutorial. I could expand it, but I'm already a bit worried about how long it takes to run the notebook (4.5 minutes on my machine, and it looks as if there is an upper limit of 10 minutes).

jpacold commented:
> Thanks, @jpacold. Regarding your question in the PR description, most definitely a small non-working example would help to understand what the issue is.

An example is below. I get a similar error if I replace the $R_{ZZ}$ gate with an $R_{XX}$ gate.

<details>
<summary>Script</summary>

```python
from qiskit import QuantumCircuit
from qiskit_aer import QasmSimulator

from mitiq.zne import mitigate_executor
from mitiq.zne.inference import RichardsonFactory

def create_circuit() -> QuantumCircuit:
    circuit = QuantumCircuit(2)
    circuit.rzz(0.1, 0, 1)
    circuit.rx(0.01, circuit.qubits)
    return circuit

def executor_with_noise(circuit: QuantumCircuit, n_shots: int = 8000) -> float:
    backend = QasmSimulator()
    backend.run(circuit, shots=n_shots)
    # return dummy value
    return 1.0

if __name__=="__main__":
    zne_executor = mitigate_executor(executor_with_noise, factory=RichardsonFactory([1.0, 2.0, 3.0]))
    result = zne_executor(create_circuit())
```
</details>

<details>
<summary>Error message</summary>

```
Traceback (most recent call last):
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/mitiq_qiskit/conversions.py", line 253, in from_qiskit
    mitiq_circuit = from_qasm(qasm2.dumps(circuit))
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/mitiq_qiskit/conversions.py", line 273, in from_qasm
    return circuit_from_qasm(qasm)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/qasm.py", line 29, in circuit_from_qasm
    return QasmParser().parse(qasm).circuit
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/_parser.py", line 538, in parse
    self.parsedQasm = self.parser.parse(lexer=self.lexer)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/ply/yacc.py", line 1120, in parseopt_notrack
    p.callable(pslice)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/_parser.py", line 340, in p_gate_op_with_params
    self._resolve_gate_operation(args=p[5], gate=p[1], p=p, params=p[3])
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/_parser.py", line 349, in _resolve_gate_operation
    raise QasmException(msg)
cirq.contrib.qasm_import.exception.QasmException: Unknown gate "rzz" at line 4

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/conversions.py", line 129, in convert_to_mitiq
    mitiq_circuit = conversion_function(circuit)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/mitiq_qiskit/conversions.py", line 259, in from_qiskit
    mitiq_circuit = from_qasm(qasm2.dumps(circuit))
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/mitiq_qiskit/conversions.py", line 273, in from_qasm
    return circuit_from_qasm(qasm)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/qasm.py", line 29, in circuit_from_qasm
    return QasmParser().parse(qasm).circuit
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/_parser.py", line 538, in parse
    self.parsedQasm = self.parser.parse(lexer=self.lexer)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/ply/yacc.py", line 1201, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/ply/yacc.py", line 192, in call_errorfunc
    r = errorfunc(token)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/cirq/contrib/qasm_import/_parser.py", line 521, in p_error
    raise QasmException(
cirq.contrib.qasm_import.exception.QasmException: Syntax error: '('
...ate r(param0,param1) q0 { u3(param0,param1 - pi/2,pi/2 - param1) q0; }
qreg q[2];
cx q[0],q[1];
rz(0.1) q[1];
cx q[0],q[1];
r(0.01,0) q[0];
r(0.01,0) q[1]
        ^
at line 3, column 7

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jpacold/mitiq/conversion_bug.py", line 21, in <module>
    result = zne_executor(create_circuit())
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/zne/zne.py", line 146, in new_executor
    return execute_with_zne(
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/zne/zne.py", line 109, in execute_with_zne
    return factory.run(
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/zne/inference.py", line 538, in run
    _check_circuit_length(qp)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/conversions.py", line 209, in accept_any_qprogram_function
    cirq_circuit, _ = convert_to_mitiq(circuit)
  File "/home/jpacold/mitiq/env/lib/python3.10/site-packages/mitiq/interface/conversions.py", line 131, in convert_to_mitiq
    raise CircuitConversionError(
mitiq.interface.conversions.CircuitConversionError: Circuit could not be converted to an internal Mitiq circuit. This may be because the circuit contains custom gates or Pragmas (pyQuil). If you think this is a bug or that this circuit should be supported, you can open an issue at https://github.com/unitaryfund/mitiq. 

 Provided circuit has type <class 'qiskit.circuit.quantumcircuit.QuantumCircuit'> and is:

               ┌──────────┐
q_0: ─■────────┤ Rx(0.01) ├
      │ZZ(0.1) ├──────────┤
q_1: ─■────────┤ Rx(0.01) ├
               └──────────┘

Circuit types supported by Mitiq are 
braket, cirq, pennylane, pyquil, qibo, qiskit.
```

</details>

cosenal commented:
Let me give you a quick explanation of the issue, and then some possible solutions.

At their core, Mitiq's mitigation techniques are implemented on circuits represented in Cirq. This means that if you pass a circuit from one of the supported frontend (Qiskit in your case), it must first be converted to Cirq. In particular, the conversion Qiskit --> Cirq happens in two steps: Qiskit --> QASM; QASM --> Cirq. For the latter step, we use Cirq's own QASM parser, and in there  $R_{ZZ}$ / $R_{XX}$ gates [are not supported](https://github.com/quantumlib/Cirq/blob/81f66b99ba08c0edcb29799eaf0c4e39aa73cce3/cirq-core/cirq/contrib/qasm_import/_parser.py#L185) yet.

You can either file this issue under the Cirq repository, and see if it makes sense for them to support those gates, or use a decomposition of those gates onto a more standard gate sets (e.g., with CNOT and RX/RZ gates).

cosenal commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,396 @@
+  language: python
+  name: python3
+---
+
```
You can add tags to the tutorial. For an example, see https://github.com/unitaryfund/mitiq/blob/36e02854e6d92c75eccc0881b51e8ea9f18e8de2/docs/source/examples/calibration-tutorial.md?plain=1#L15

jpacold replied:
Done in cbd9028.

cosenal commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,396 @@
+```
+
+## Error mitigation with zero-noise extrapolation
+At this level of noise, we can use ZNE to mostly recover the ideal result. Running the noisy simulation is expensive (especially as the scale factor $\alpha$ increases), so rather than using the high-level functions for ZNE, we apply the static method `RichardsonFactory.extrapolate` to the results from the previous cell.
````
@natestemen is there a mechanism here to make `RichardsonFactory.extrapolate` a cross-link to the same function in the API docs?

natestemen replied:
Yes! You can use following syntax to have the inline code _also_ be a link to the API-doc.
```md
... {meth}`.RichardsonFactory.extrapolate` ...
```

More details about cross-referencing in sphinx can be found [here](https://www.sphinx-doc.org/en/master/usage/referencing.html
).

jpacold replied:
Done (on the second attempt :smile:) in 9cf927e

natestemen replied:
Ahh apologies if my comment was confusing!

cosenal commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,396 @@
+```
+
+```{code-cell} ipython3
+result_zne = RichardsonFactory.extrapolate(scale_factors,
+                                           [r[1] for r in result_depolarizing_scaled])
````
nit: we don't have strict coding style guidelines in Mitiq, but especially in tutorials, I am not a fan of this identation, because part of the code ends up being hidden at the right of the cell. Imo the following is more readable:

```suggestion
result_zne = RichardsonFactory.extrapolate(
    scale_factors,
    [r[1] for r in result_depolarizing_scaled]
)
```

jpacold replied:
Done in 454a6be

cosenal reviewed (commented):
I gave it a first review pass, very nice tutorial!

I will try to run it on my own as well, and see if there are ways to speed it up.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,396 @@
+```
+
+```{code-cell} ipython3
+result_depolarizing = run_depolarizing_noise(noise_level=0.0005)
````
Done in 4ec74e5.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,396 @@
+add_to_plot(result_depolarizing_scaled[0][0], result_zne, "mitigated", legend)
+plt.legend(legend)
+plt.show()
+```
````
Added a short summary in b72a6db.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,396 @@
+
+# ZNE with Qiskit: Simulation of Loschmidt Echo Revival
+
+This tutorial replicates some of the results from Y. Javanmard et al., ["Quantum simulation of dynamical phase transitions in noisy quantum devices"](https://arxiv.org/abs/2211.08318). We build a circuit that simulates the time-evolution of a transverse-field Ising model, then run ideal, noisy, and noise-mitigated simulations of the circuit.
```
I expanded the introduction some more in 99052e3 and 9cf927e.

natestemen commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+from qiskit.compiler import transpile
+from qiskit_aer.noise import NoiseModel
+from qiskit_aer.noise.errors.standard_errors import depolarizing_error
+from mitiq.zne.scaling.folding import fold_gates_at_random
```
```suggestion
from mitiq.zne.scaling import fold_gates_at_random
```

natestemen commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+    
+    job = backend.run(folded_circuit, shots=n_shots)
+
+    # Get the probability of returning to $\ket{0^{\otimes N}}$
```
Same nit as above.
```suggestion
    # Get the probability of returning to |0^N>
```

natestemen commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+setup_plot("Scaled noisy simulations")
+legend = []
+add_to_plot(*result_ideal, "ideal", legend)
+for alpha, result in zip(scale_factors, result_depolarizing_scaled):
+    add_to_plot(*result, "$\\alpha = {}$".format(alpha), legend)
```
It's a little more common to use $\lambda$ as the scale factor, but it looks like $\alpha$ is being used in the paper you're pulling from, so lets stick with that for now.

Can you add a note _before_ the plot indicating what $\alpha$ is? I think it'll help someone who is skimming understand the plots quicker.

jpacold replied:
Done in d8099eb.

natestemen commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,428 @@
+```
+
+## Error mitigation with zero-noise extrapolation
+At this level of noise, we can use ZNE to mostly recover the ideal result. Running the noisy simulation is expensive (especially as the scale factor $\alpha$ increases), so rather than using the high-level functions for ZNE, we apply the static method {meth}`.RichardsonFactory.extrapolate` to the results from the previous cell.
````
Smart! And great to demonstrate this workflow since I don't think we walk through this as nicely as you do here.

natestemen commented on docs/source/examples/loschmidt_echo_revival_zne.md:
````diff
@@ -0,0 +1,428 @@
+Increasing the baseline noise level makes it much harder to reconstruct the peak with ZNE.
+
+```{code-cell} ipython3
+high_noise = 0.005
+scale_factors = [1, 2, 3]
+result_depolarizing_scaled = [
+    run_depolarizing_noise(
+        noise_level=high_noise,
+        scale_factor=alpha)
+    for alpha in scale_factors
+]
+```
+
+```{code-cell} ipython3
+result_zne = RichardsonFactory.extrapolate(
+    scale_factors,
+    [r[1] for r in result_depolarizing_scaled]
+)
````
Lets merge these two cells.

jpacold replied:
Done in 8d3acee.

natestemen reviewed (commented):
What a great tutorial @jpacold! I learned a lot from reviewing it. I left some minor requests, but nothing that should require much work.

Thank you for contributing this tutorial, and well done recreating the results from the paper :)

cosenal commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+    
+    psi = job.result().data()["statevector"]
+    
+    # Get the probability of returning to $\ket{0^{\otimes N}}$
```
or `|00...0>`

jpacold replied:
Done in 8d3acee.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+
+# ZNE with Qiskit: Simulation of Loschmidt Echo Revival
+
+This tutorial replicates some of the results from Y. Javanmard et al., ["Quantum simulation of dynamical phase transitions in noisy quantum devices"](https://arxiv.org/abs/2211.08318). We build a circuit that simulates the time-evolution of a transverse-field Ising model, then run ideal, noisy, and noise-mitigated simulations of the circuit.
```
Thanks a lot for all the feedback -- did this in 5f31850.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+
+The paper considers some additional effects of noise, which are outside the scope of this tutorial:
+   
+* Let $\lambda(t) = \lim_{N->\infty} -\log(\Lambda(t))/N$, where $N$ is the number of sites in the Ising model. Dynamical quantum phase transitions (DQPTs) occur at values of $t$ where $\lambda(t)$ is not analytic. DQPTs are observed at different times in the ideal and noisy simulations, and occur more frequently in the noisy system.
```
Fixed the typo in 5f31850.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+
+$$H_{zz} = -\frac{1}{2} \left[ \sum_{i=0}^{N-2}J_z Z_i Z_{i+1} \right], \hspace{0.4cm} H_{xx} = -\frac{1}{2} \left[ \sum_{i=0}^{N-2}J_x X_{i} X_{i+1} \right], \hspace{0.4cm} H_x = -\frac{1}{2} \left[ \sum_{i=0}^{N-1} h_x X_i \right]$$
+
+where $X_i$ and $Z_i$ are the Pauli operators acting on site $i$, $J_z$ and $J_x$ are the $z$- and $x$-components of the spin-spin coupling, and $h_x$ is the strength of the external field. (Strictly speaking, when $J_x \neq 0$ this is a Heisenberg model rather than an Ising model.)
```
I added a short note on this in dfa3503. In the textbook Ising model all the constants are real. There are systems where one or more of the constants is effectively imaginary but to be honest I'm not very familiar with the literature on them.

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,428 @@
+
+$$\exp(-i H_{zz} \delta t) = \prod_{i=0}^{N-2} \exp\left( -i\frac{J_z\delta t }{2} Z_i Z_{i+1} \right)$$
+
+Using the fact that $Z_i Z_{i+1} = I \otimes \cdots Z \otimes Z \cdots \otimes I$, we can rewrite this as a product of $R_{ZZ}$ gates,
```
Added links in 8d3acee.

cosenal reviewed (approved):
Looks great!
@natestemen I added this to the 0.41 milestone

natestemen reviewed (approved):
Really great work @jpacold! This will be included in the next release of Mitiq which will happen tomorrow. I'll update you with a live link then :)

By the way, if you want to do more with Mitiq we're happy to help! Feel free to make feature requests, ask questions, and you're always welcome at our community call on Fridays which takes place on our [discord](http://discord.unitary.fund).

jpacold commented on docs/source/examples/loschmidt_echo_revival_zne.md:
```diff
@@ -0,0 +1,425 @@
+    t_max : float = 8.5
+    M : int = 25
+    n_qubits : int = 6
+    dt : float = dataclasses.field(init=False)
+    t_vals : np.ndarray[float] = dataclasses.field(init=False)
```
AFAIK the only thing it does here is remove `dt` and `t_vals` from the arguments of `SimulationsParams.__init__`, so it's a little clearer that we want to calculate them from the other parameters.
